libdrgn/debug_info.c

```diff
@@ -2443,8 +2444,17 @@ drgn_type_from_dwarf_internal(struct drgn_debug_info *dbinfo,
 			if (!dwarf)
 				return drgn_error_libdwfl();
 			uintptr_t start = (uintptr_t)module->scns[DRGN_SCN_DEBUG_INFO]->d_buf;
-			if (!dwarf_offdie(dwarf, die_addr - start, die))
-				return drgn_error_libdw();
+			uint64_t size = module->scns[DRGN_SCN_DEBUG_INFO]->d_size;
+			if (die_addr >= start &&
```

Comment:
I added a bit of duplication here, but I can make a helper in dwarf_index.h if you'd like to combine them.